rfcs/20250911-mxfp-support/README.md

```diff
@@ -0,0 +1,224 @@
+`dynamic_scales` attribute would be mutually exclusive. Hence only one
+quantization scheme would be accepted for a given memory argument.
+Also, this allows to reuse `DNNL_ARG_ATTR_SCALES` argument kind mask
+to pass dynamic scales buffer to the `execute` function.
+
+If not mutually exclusive, we would need to expose a new argument kind
+`DNNL_ARG_ATTR_DYNAMIC_SCALES` for users to specify the memory
+argument for writing computed scales.
+
+To summarize, here is an example of this new API usage with mxfp4:
+``` C++
+    // Elements type for all in/out puts are e2m1 for MXFP4
+    memory::desc a_md({M, K}, memory::data_type::f4_e2m1, {K, 1}); // M x K layout
+    memory::desc b_md({K, N}, memory::data_type::f4_e2m1, {K, 1}); // N x K layout
+    memory::desc c_md({M, N}, memory::data_type::f4_e2m1, {N, 1}); // M x N layout
+
+    // Create attributes and set static scales for inputs, 
+    // with type of e8m0 and groups of 32 along K. 
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_SRC,
+            /* mask */ (1 << 0) + (1 << 1), {1, 32}, memory::data_type::e8m0);
+    attr.set_scales(DNNL_ARG_WEIGHTS,
+            /* mask */ (1 << 0) + (1 << 1), {32, 1}, memory::data_type::e8m0);
+
+    // Set dynamic scales for output, with type of e8m0 and groups of 32 along K.
+    attr.set_dynamic_scales(DNNL_ARG_DST,
+            /* mask */ (1 << 0) + (1 << 1), {1, 32}, memory::data_type::e8m0);
+
+    // Create a MatMul primitive descriptor
+    matmul::primitive_desc(eng, a_md, b_md, c_md, attr);
+```
+
+### Option 1.b (recommended): extend scales attribute (`set_scales`)
+
+Same principle as option 1.a. except that instead of exposing new
+attribute, `set_scales` would be extended to support dynamic
+scaling. Main benefit here would be that:
+- the API itself would make it clear that static and dynamic
+quantization are mutually exclusive.
+- it would allow to extend to other quantization formulas like static
+  quantization with floating point zero-point as $`x_{f32} =
+  scale*x_q - zp`$ instead of $`x_{f32} = scale * (x_q - zp)`$, or
+  dynamic quantization formulas where conversion happens after
+  division.
+
+We would here introduce a new enum class and add a new argument to
+set_scales.  For C API, this would require a v3 method as follow:
+
+```C
+typedef enum {
+    dnnl_quantization_kind_undef = 0,
+    dnnl_quantization_kind_static,
+    dnnl_quantization_kind_dynamic_mx,
+
+} dnnl_quantization_kind_t;
+
+dnnl_status_t DNNL_API dnnl_primitive_attr_set_scales_v3(
+        dnnl_primitive_attr_t attr, int arg, int mask, int ndims,
+        const dnnl_dims_t group_dims, dnnl_data_type_t data_type,
+        int is_on_host, dnnl_quantization_kind qkind);
+```
+
+In C++ API, we would introduce similar enum type and extension to `set_scales`
+
+```C++
+enum class quantization_kind {
+    undef = dnnl_quantization_kind_undef,
+    static_ = dnnl_quantization_kind_static,
+    dynamic_mx = dnnl_quantization_kind_dynamic_mx,
```

Comment:
I wonder if `dynamic_mx` is precise enough to describe the intent when it comes to fp4 data type and if it requires more words in the name to distinct it from a similar but different flavor.

Reply:
It should be independent of elements datatype. Here it is about OCP MX spec definition of how scales are computed.